README.md
# ASJOverflowButton

Working on projects that are simultanously being developed on both iOS and Android can post interesting challenges, especially concerning the look and feel of both apps. There are standard controls on both platforms that are not available in the other. [Overflow menu](https://developer.android.com/design/patterns/actionbar.html) is one of them. This library is a `UIBarButtonItem` subclass that you can put on a navigation bar and an overflow menu will be presented on its tap.

# Installation

CocoaPods is the preferred way to install this library. Add this command to your `Podfile`:

```ruby
pod 'ASJOverflowButton'
```

# Usage

Use the designated initializer to create the button, like so:

```objc
ASJOverflowButton *overflowButton = [[ASJOverflowButton alloc] initWithTarget:aController title:@"An optional title" image:anOptionalImage button:anOptionalButton items:_overflowItems];
self.navigationItem.rightBarButtonItem = overflowButton;
```
The items need to be on type `ASJOverflowItem`. Constructor method is provided to generate them. Just attach an array of these during initialization.

```objc
ASJOverflowItem *item = [ASJOverflowItem itemWithName:itemName image:image backgroundColor:color];
```

The are a few properties to customize the look and feel of the overflow menu.

```objc
@property (nullable, strong, nonatomic) UIColor *menuBackgroundColor;
```
Sets the background color of the overflow menu. Defaults to white.

```objc
@property (nullable, strong, nonatomic) UIColor *itemTextColor;
```
Sets the text color of the menu items. Defaults to black.

```objc
@property (nullable, strong, nonatomic) UIColor *itemHighlightedColor;
```
Sets the menu item's background color when tapped. Defaults to RGB (217, 217, 217).

```objc
@property (nullable, strong, nonatomic) UIColor *borderColor;
```
Sets the color of border for the menu. Defaults to gray.

```objc
@property (assign, nonatomic) CGFloat borderWidth;
```
Sets the width of the menu's border. Defaults to zero.

```objc
@property (nullable, strong, nonatomic) UIFont *itemFont;
```
Sets the font of the menu items. Defaults to system font of size 17 pts.

```objc
@property (assign, nonatomic) BOOL hidesSeparator;
```
Hides the separator between two menu items. Needs to be set `NO` for `separatorInsets` property to work. Defaults to `YES`.

```objc
@property (assign, nonatomic) BOOL hidesShadow;
```
Hides the shadow around the menu. Defaults to `NO`.

```objc
@property (assign, nonatomic) BOOL dimsBackground;
```
Dims the background when menu is showing. Defaults to `NO`.

```objc
@property (assign, nonatomic) CGFloat dimmingLevel;
```
Sets the degree to which the background is dimmed when menu is shown. Will work only if `shouldDimBackground ` is set to `YES`. Ranges from 0.0 to 1.0. Defaults to 0.6.

```objc
@property (assign, nonatomic) CGFloat menuItemHeight;
```
Sets the height of individual overflow menu items. Defaults to 40 pts.

```objc
@property (assign, nonatomic) CGFloat widthMultiplier;
```
A ratio to set the menu width with respect to the screen width. Acceptable values are from 0.0 to 1.0. For example, if your screen width is 320.0 pts and 'widthMultiplier' is set to 0.5, the menu width will be 0.5 * 320.0 = 160 pts. Defaults to 0.4.

```objc
@property (assign, nonatomic) SeparatorInsets separatorInsets;
```
Sets the left and right insets of the separator between two menu items. Works only if `hidesSeparator` is set to 'NO'. Defaults to (15.0f, 0.0f).

```objc
@property (assign, nonatomic) MenuMargins menuMargins;
```
Sets the margins of the menu from the top, right and bottom edges of the screen. Menu will always appear below the status bar. Defaults to 5 pts each.

```objc
@property (assign, nonatomic) MenuAnimationType menuAnimationType;
```
Sets the way the menu should be shown. There are two options, fading it in or zooming in from the top right corner. Defaults to `MenuAnimationTypeZoomIn`.

![alt tag](Images/Normal.png)
![alt tag](Images/Dimmed.png)

### Limitation

Can only show the overflow menu on the top right corner of the screen. From [release 1.2](https://github.com/sudeepjaiswal/ASJOverflowButton/releases/tag/1.2) however, you can adjust the `menuMargins` property to change the menu position.

# To-do

- ~~Menu closes abruptly when tapped outside~~
- ~~Menu should close when tapped on the clear table view outside the items~~
- ~~Allow a menu to be created with a title~~
- ~~Some way to adjust width of the menu~~
- Maybe some way to show menu on both, left and right sides
- Gradient at bottom to hint scrolling

# Credits

- To [Shashank Pali](https://github.com/shashankpali) to make the menu zoom in code work - Read a [blog post](http://shashankpali.weebly.com/posts/scale-view-from-anywhere-cgaffinetransform) about it
- To [Ken M. Haggerty](https://github.com/kenmhaggerty) for adding new features and logic
- To [Vishal Deshai](https://github.com/vishaldeshai) for helping fix RTL issue
- to [Fuzzball](https://github.com/fuzzball1980) for helping implement custom buttons
- [UITableView passes touch events to superview when it shouldn't](http://stackoverflow.com/questions/11570160/uitableview-passes-touch-events-to-superview-when-it-shouldnt)
- [Can i change multiplier property for NSLayoutConstraint?](http://stackoverflow.com/questions/19593641/can-i-change-multiplier-property-for-nslayoutconstraint#comment54574185_27831003)
- [Which is the best way to suppress “unused variable” warning?](http://stackoverflow.com/questions/17622237/which-is-the-best-way-to-suppress-unused-variable-warning)

# License

`ASJOverflowButton` is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## [1.6](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.6)
Released on Wednesday, 26 Jan, 2022.

#### Added
* Added `borderColor` and `borderColor` properties.
* Added delegate for menu and block for individual menu items.

#### Fixed
* [Issue #3](https://github.com/sdpjswl/ASJOverflowButton/issues/3) where already open menu was getting displaced upon device rotation.
* Menu still opening on right side for for RTL regions.
* Crash while opening xib from wrong bundle.

#### Updated
* The designated initializer to optionally pass title or custom button in addition to button image.

## [1.5.2](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.5.2)
Released on Wednesday, 2 Nov, 2016.

#### Added
* Added `backgroundColor` property to `ASJOverflowItem` to set color to individual items.

## [1.5.1](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.5.1)
Released on Wednesday, 2 Nov, 2016.

#### Fixed
* Partly fixed [issue #3](https://github.com/sdpjswl/ASJOverflowButton/issues/3). Menu opens correctly if opened in landscape orientation.
* Fixed UI bug while scrolling the menu if only some items have image.

## [1.5](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.5)
Released on Saturday, 1 Oct, 2016.

#### Fixed
* Fixed [issue #2](https://github.com/sdpjswl/ASJOverflowButton/issues/2). Overflow menu shadow drawing incorrectly on iOS 10.

## [1.4](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.4)
Released on Tuesday, 16 Aug, 2016.

#### Added
* Added new properties 'hidesSeparator' and 'separatorInsets' to control the visibility and position of the menu item separators.

## [1.3.1](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.3.1)
Released on Thursday, 4 Aug, 2016.

#### Updated
* Updated instances throughout the project and readme where 'dimsBackground' was appearing as its old name 'shouldDimBackground'.

## [1.3](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.3)
Released on Wednesday, 3 Aug, 2016.

#### Added
* Added new property 'dimmingLevel' to control the degree to which the background can be dimmed when the menu is shown.
* Added new property 'menuAnimationType' to change the way the menu is shown; fade in or zoom in.

## [1.2](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.2)
Released on Monday, 18 Jul, 2016.

#### Added
* Added new property 'widthMultiplier' to set menu width ratio with respect to screen width.
* Added new property 'menuMargins' to set menu margins from the top, right and bottom edges of the screen.

#### Updated
* Updated readme to reflect new properties and added more documentation.

#### Fixed
* Fixed/suppressed release warnings.

## [1.1](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.1)
Released on Tuesday, 12 Jul, 2016.

#### Added
* Added a constructor method to create overflow item with only title.

#### Updated
* Updated the designated initializer; removed target controller. Menu is now created on a separate window.

#### Fixed
* Fixed a bug where shadow wasn't being drawn properly.

## [1.0](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/1.0)
Released on Monday, 11 Jul, 2016.

#### Added
* Added change log.

#### Fixed
* Fixed a bug where menu wasn't dismissed if tap happened below the menu items.

## [0.2](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/0.2)
Released on Tuesday, 17 May, 2016.

#### Added
* Added documentation.
* Added new initializer.

## [0.1](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/0.1)
Released on Wednesday, 11 May, 2016.

#### Added
* Added nullability annotations.
* Merged pull request #1.

## [0.0.1](https://github.com/sdpjswl/ASJOverflowButton/releases/tag/0.0.1)
Released on Tuesday, 2 Feb, 2016.

#### Added
* Initial release.


ASJOverflowButton/ASJOverflowButton.h
//
// ASJOverflowButton.h
//
// Copyright (c) 2015-2016 Sudeep Jaiswal
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


#import "ASJOverflowMenu.h"
#import <UIKit/UIBarButtonItem.h>

@class ASJOverflowItem;

NS_ASSUME_NONNULL_BEGIN

typedef void (^ItemTapBlock)(ASJOverflowItem *item, NSInteger idx);
typedef void (^HideMenuBlock)(void);

@interface ASJOverflowButton : UIBarButtonItem

/**
 *  Called in response to menu item selection. Optional.
 */
@property (nullable, weak, nonatomic) id <ASJOverflowMenuDelegate> delegate;

/**
 *  The overflow menu's background color. Defaults to white.
 */
@property (nullable, strong, nonatomic) UIColor *menuBackgroundColor;

/**
 *  The overflow menu items' text color. Defaults to black.
 */
@property (nullable, strong, nonatomic) UIColor *itemTextColor;

/**
 *  The selected overflow menu item's background color when tapped. Defaults to RGB (217, 217, 217).
 */
@property (nullable, strong, nonatomic) UIColor *itemHighlightedColor;

/**
 *  The overflow menu's border color. Defaults to gray.
 */
@property (nullable, strong, nonatomic) UIColor *borderColor;

/**
 *  The overflow menu's border width. Defaults to zero.
 */
@property (assign, nonatomic) CGFloat borderWidth;

/**
 *  The overflow menu items' font. Defaults to system font 17 pts.
 */
@property (nullable, strong, nonatomic) UIFont *itemFont;

/**
 *  It set YES, the separator between two menu items will be hidden. In that case, setting 'separatorInsets' will do nothing. Defaults to 'YES'.
 */
@property (assign, nonatomic) BOOL hidesSeparator;

/**
 *  If set YES, the shadow around the menu will not be drawn. Defaults to 'NO'.
 */
@property (assign, nonatomic) BOOL hidesShadow;

/**
 *  If set YES, the background will be dimmed while the menu is visible. Defaults to 'NO'.
 */
@property (assign, nonatomic) BOOL dimsBackground;

/**
 *  Sets the degree to which the background is dimmed when menu is shown. Will work only if 'dimsBackground' is set to 'YES'. Ranges from 0.0 to 1.0. Defaults to 0.6.
 */
@property (assign, nonatomic) CGFloat dimmingLevel;

/**
 *  Sets the height of individual overflow menu items. Defaults to 40 pts.
 */
@property (assign, nonatomic) CGFloat menuItemHeight;

/**
 *  Sets the ratio according to which the overflow menu width should be calculated. Acceptable values are from 0.0 to 1.0. For example, if your screen width is 320.0 pts and 'widthMultiplier' is set to 0.5, the menu width will be 0.5 * 320.0 = 160 pts. Defaults to 0.4.
 */
@property (assign, nonatomic) CGFloat widthMultiplier;

/**
 *  Sets the left and right insets of the separator between two menu items. Works only if 'hidesSeparator' is set to 'NO'. Defaults to (15.0f, 0.0f).
 */
@property (assign, nonatomic) SeparatorInsets separatorInsets;

/**
 *  The margins of the menu from the top, right and bottom edges. Defaults to 5 pts each.
 */
@property (assign, nonatomic) MenuMargins menuMargins;

/**
 *  Sets which way the menu should be shown; options including fading it in or zooming in from the top right corner. Defaults to 'MenuAnimationTypeZoomIn'.
 */
@property (assign, nonatomic) MenuAnimationType menuAnimationType;

/**
 *  A block that is called when any overflow menu item is tapped.
 */
@property (nullable, copy) ItemTapBlock itemTapBlock;

/**
 *  A block that is called when any overflow menu is removed from the screen.
 */
@property (nullable, copy) HideMenuBlock hideMenuBlock;

/**
 *  The designated initializer. You can use either title, image or button. Just pass nil for what you don't require.
 *
 *  @param target A UIViewController to show the overflow menu on.
 *  @param title  The overflow button's title. Used if image and button are not provided. Defaults to "Open",
 *  @param image  The overflow button's image.
 *  @param button A custom UIButton to show on the navigation bar.
 *  @param items  An array of ASJOverflowItems to show on the menu.
 *
 *  @return An instance of ASJOverflowButton.
 */
- (instancetype)initWithTarget:(__kindof UIViewController *)target title:(nullable NSString *)title image:(nullable UIImage *)image button:(nullable UIButton *)button items:(NSArray<ASJOverflowItem *> *)items NS_DESIGNATED_INITIALIZER;

/**
 *  Don't allow user to use "init".
 */
- (instancetype)init NS_UNAVAILABLE;

/**
 *  Don't allow user to use "initWithCoder:".
 */
- (instancetype)initWithCoder:(NSCoder *)aDecoder NS_UNAVAILABLE;

- (void)setItemTapBlock:(ItemTapBlock _Nullable)itemTapBlock;
- (void)setHideMenuBlock:(HideMenuBlock _Nullable)hideMenuBlock;

@end

@interface ASJOverflowItem : NSObject

/**
 *  The overflow item's name.
 */
@property (copy, nonatomic) NSString *name;

/**
 *  The overflow item's image. Optional.
 */
@property (nullable, strong, nonatomic) UIImage *image;

/**
 *  The overflow item's background color. Optional.
 */
@property (nullable, strong, nonatomic) UIColor *backgroundColor;

/**
 *  A block that is called when the overflow menu item is tapped.
 */
@property (nullable, strong, nonatomic) ItemTapBlock itemTapBlock;

/**
 *  A convenience constructor to create ASJOverflowItems.
 *
 *  @param name  The overflow item's name.
 *
 *  @return An instance of ASJOverflowItem.
 */
+ (ASJOverflowItem *)itemWithName:(NSString *)name;

/**
 *  A convenience constructor to create ASJOverflowItems.
 *
 *  @param name  The overflow item's name.
 *  @param itemTapBlock  A block that gets called when the overflow item is tapped. Optional.
 *
 *  @return An instance of ASJOverflowItem.
 */
+ (ASJOverflowItem *)itemWithName:(NSString *)name tapBlock:(ItemTapBlock)itemTapBlock;

/**
 *  A convenience constructor to create ASJOverflowItems.
 *
 *  @param name  The overflow item's name.
 *  @param image The overflow item's image. Optional.
 *  @param backgroundColor The overflow item's background color. Optional.
 *
 *  @return An instance of ASJOverflowItem.
 */
+ (ASJOverflowItem *)itemWithName:(NSString *)name image:(nullable UIImage *)image backgroundColor:(nullable UIColor *)backgroundColor;

@end

NS_ASSUME_NONNULL_END
